play/tests/test_jamesk.py

```diff
@@ -0,0 +1,25 @@
+import pytest
+
+from play.jamesk import knockknock
+
+def test_knockknock():
+    import sys
+    import io
+    import re
+
+    # redirect stdout to a string stream
+    stream = io.StringIO()
+    sys.stdout = stream
+    # exercise the function
+    knockknock()
+    # restore the stdout
+    sys.stdout = sys.__stdout__
+    # evaluate the results
+    full_joke = stream.getvalue()
+    stream.close()
+    pattern = re.compile(r"Knock Knock...\n\t\tWho's there\?\n([^\n]+)\n\t\t([\w]+) who\?\n([^\n]+)\n")
```

Comment:
Mostly from fear of the unknown of Windows vs. mac/linux newlines (`\r\n` vs `\n`), I try to avoid hard-coded newlines when possible. I was initially surprised to see that this works on Windows, but from what I can see Python emits `\n` when printing but accepts `\r\n` or `\n` as new lines. 

Anyway, I would have split `full_joke` into lines with `.splitlines()` and then checked each one separately.

Reply:
I decided to update the regular expression to handle the possibility of `\r\n` instead of splitting the lines.  I explored splitting it, but the for loop or multiple checks looked like overkill compared to the regular expression matching.

Reply:
I have a pretty low success rate with getting complicated regex's right, so I'll write a few extra lines if I can avoid them. E.g. I'd probably do something like:
```
lines = full_joke.splitlines()
assert len(lines) == 5
assert lines[0] == 'Knock Knock...'
assert lines[1] == "\t\tWho's there?"
assert lines[2] in knockknock.jokes
assert lines[3] == f'\t\t{lines[2]} who?'
assert lines[4] == knockknock.jokes[lines[2]]
```
But if you love regex, then using verbose mode helps alot:
```
regex = re.compile(
    r'Knock \s Knock... \r?\n'
    r'\t\t Who's \s there \?\r?\n'
    r'([^\r\n]+) \r?\n'
    r'\t\t ([\w]+) \s who\? \r?\n'
    r'([^\r\n]+) \r?\n', 
    re.VERBOSE)
```